### PR TITLE
_cli, setup: experimental tab completion generation

### PR DESCRIPTION
**Pull request by woodruffw**
This is a **demonstration** of what automatic tab completion generation for `pip-audit` might look like. It uses [shtab](https://github.com/iterative/shtab), which does the generation under the hood without requiring more invasive system hooks.

Example use (where `<tab>` indicates a physical tab):

```bash
$ eval "$(pip-audit --completions=bash)"

$ pip-audit -<tab>
--cache-dir              --progress-spinner       -f
--completions            --requirement            -h
--desc                   --timeout                -l
--dry-run                --version                -r
--format                 --vulnerability-service  -s
--help                   -V
--local                  -d

$ pip-audit --vulnerability-service<tab>
osv   pypi
```

This approach seems relatively extensible (`shtab` seems to support completion extensions, like indicating that an option takes filenames only) as well.

### CODE REVIEW

**Comment by woodruffw** [comment]
Upstream type hints request: https://github.com/iterative/shtab/issues/59

**Review by casperdcl — APPROVED** [review]

**Comment by tetsuo-cpp** [comment]
I seem to be having trouble using this with the default `zsh` on my Mac.
```sh
(env) tetsuo@Alexs-MacBook-Pro pip-audit % $SHELL --version
zsh 5.8 (x86_64-apple-darwin21.0)
(env) tetsuo@Alexs-MacBook-Pro pip-audit % pip-audit --completions=zsh
Traceback (most recent call last):
  File "/Users/tetsuo/Code/pip-audit/env/bin/pip-audit", line 33, in <module>
    sys.exit(load_entry_point('pip-audit', 'console_scripts', 'pip-audit')())
  File "/Users/tetsuo/Code/pip-audit/pip_audit/_cli.py", line 203, in audit
    print(shtab.complete(parser, shell=args.completions))
  File "/Users/tetsuo/Code/pip-audit/env/lib/python3.9/site-packages/shtab/__init__.py", line 705, in complete
    return completer(
  File "/Users/tetsuo/Code/pip-audit/env/lib/python3.9/site-packages/shtab/__init__.py", line 568, in complete_zsh
    root_options="\n  ".join(
  File "/Users/tetsuo/Code/pip-audit/env/lib/python3.9/site-packages/shtab/__init__.py", line 569, in <genexpr>
    format_optional(opt) for opt in parser._get_optional_actions()
  File "/Users/tetsuo/Code/pip-audit/env/lib/python3.9/site-packages/shtab/__init__.py", line 472, in format_optional
    (choice_type2fn[opt.choices[0].type] if isinstance(opt.choices[0], Choice) else
  File "/opt/homebrew/Cellar/python@3.9/3.9.7_1/Frameworks/Python.framework/Versions/3.9/lib/python3.9/enum.py", line 432, in __getitem__
    return cls._member_map_[name]
KeyError: 0
(env) tetsuo@Alexs-MacBook-Pro pip-audit % $SHELL --version
zsh 5.8 (x86_64-apple-darwin21.0)
```
Not sure if it's just PEBCAK on my part, or whether there's something to solve here. If I get time, I'll take a look at this before I sign off tonight.

**Comment by di** [comment]
I was -1 on this until I noticed that pip also has tab completion now: https://pip.pypa.io/en/stable/user_guide/#command-completion.

I'm now +0 on adding tab completion, but -1 on doing this in a way that would make it difficult to vendor this into `pip` in the future, or make the behavior subtly different in some way if it was, which means that if we want to do this, we should do it the same way `pip` does it, or not at all IMO.

**Comment by casperdcl** [comment]
> I seem to be having trouble using this with the default `zsh` on my Mac.

ah yes `enum.Enum` choices are [discouraged by argparse](https://docs.python.org/3/library/argparse.html#choices) and aren't supported by `shtab`... would either have to use a `list` or open an [upstream PR](https://github.com/iterative/shtab/pulls)

**Comment by tetsuo-cpp** [comment]
Thanks for the clarification @casperdcl.

@di 
> I'm now +0 on adding tab completion, but -1 on doing this in a way that would make it difficult to vendor this into pip in the future, or make the behavior subtly different in some way if it was, which means that if we want to do this, we should do it the same way pip does it, or not at all IMO.

The way I imagined `pip-audit` getting vendored into `pip` was that the code in `pip_audit/_cli.py` wouldn't be included since it's just glue that wires the CLI options to the public API (and it will eventually have flags for things like docker images which we wouldn't want in `pip`). So I wouldn't expect this change to make things harder, although @woodruffw would have a better idea.

**Comment by woodruffw** [comment]
`pip` does autocompletion a little cleverly:

* They generate a very small amount of upfront completion code: https://github.com/pypa/pip/blob/main/src/pip/_internal/commands/completion.py; that code then essentially causes `pip` to be called with `PIP_AUTO_COMPLETE=1` at each `<tab>` point
* They handle each `<tab>` here: https://github.com/pypa/pip/blob/df98167fe54b99f61dd000ae619fd4b1c634c039/src/pip/_internal/cli/autocompletion.py#L15-L108

This approach is very adaptable to `pip-audit` -- we'll probably end up turning it into a subcommand of `pip` by creating a new `Command` subclass (similar to all other `pip` subcommands), at which point (I think) we'll get tab completion for free.

So to summarize: the approach in this PR is probably a little overkill and will eventually be obviated by integration with `pip`. I'll close it.

Separately: @tetsuo-cpp raises a great point about how we plan to integrate the CLI. My original plan was to see whether we could re-use most of `_cli.py` by leveraging `argparse`'s ability to add subparsers. But we'll need to look further into that.

**Comment by casperdcl** [comment]
hmm, one disadvantage of the `pip` approach is the side-effect of executing the program every time someone presses `tab`. Plus also needs much more code to implement (compared to just passing an `argparse` object to an external library). Anyway, up to you.
